pennylane/tape/interfaces/functions.py

```diff
@@ -0,0 +1,198 @@
+    sum = tf.reduce_sum
+    tensordot = tf.tensordot
+    transpose = tf.transpose
+    zeros = tf.zeros
+
+    @staticmethod
+    def scatter_element_add(tensor, index, value):
+        return tensor + TensorFlowFunctions.scatter_element(index, value, tensor.shape)
+
+    @staticmethod
+    def scatter_element(index, value, new_dimensions):
+        indices = tf.expand_dims(index, 0)
+        tensor = tf.expand_dims(value, 0)
+        return TensorFlowFunctions.scatter(indices, tensor, new_dimensions)
+
+    @staticmethod
+    def block_diag(*tensors):
+        linop_blocks = [tf.linalg.LinearOperatorFullMatrix(block) for block in tensors]
+        linop_block_diagonal = tf.linalg.LinearOperatorBlockDiag(linop_blocks)
+        return linop_block_diagonal.to_dense()
+
+
+class TorchFunctions:
+    """Common wrapper functions for PyTorch"""
+
+    float64 = torch.float64
+    complex128 = torch.complex128
+
+    abs = torch.abs
+    asarray = torch.as_tensor
+    cast = lambda tensor, dtype: tensor.to(dtype=dtype)
+    diag = lambda tensor: torch.diag(torch.stack(tensor))
+    dot = torch.dot
+    einsum = torch.einsum
+    expand_dims = lambda a, axis: torch.unsqueeze(a, axis)
+    flatten = torch.flatten
+    gather = lambda array, indices: array[indices]
+    reduce_sum = torch.sum
+    reshape = torch.reshape
+    sqrt = torch.sqrt
+    stack = torch.stack
+    sum = torch.sum
+    tensordot = torch.tensordot
+    transpose = lambda tensor: tensor.T
+    zeros = torch.zeros
+
+    @staticmethod
+    def scatter_element_add(tensor, index, value):
+        tensor[tuple(index)] = value
+        return tensor
+
+    @staticmethod
+    def scatter(indices, tensor, new_dimensions):
+        new_array = torch.zeros(new_dimensions, dtype=tensor.dtype)
+        new_array[indices] = tensor
+        return new_array
+
+    @staticmethod
+    def block_diag(*tensors):
+        sizes = [t.shape[0] for t in tensors]
+        res = torch.zeros([sum(sizes), sum(sizes)], dtype=tensors[0].dtype)
+
+        p = np.cumsum(sizes)
+        indices = np.vstack([p - sizes, p]).T
+
+        for t, p in zip(tensors, indices):
+            row = np.arange(*p).reshape(-1, 1)
+            col = np.arange(*p).reshape(1, -1)
+            res[row, col] = t
+
+        return res
+
+
+MLFunctionWrapper = {
+    "np": NumPyFunctions,
+    "numpy": NumPyFunctions,
+    "autograd": NumPyFunctions,
+    "pennylane": NumPyFunctions,
+    "builtins":NumPyFunctions,
+    "tf": TensorFlowFunctions,
+    "tensorflow": TensorFlowFunctions,
+    "torch": TorchFunctions,
+}
+"""dict[str, namespace]: dictionary mapping various machine learning
+library namespaces/nicknames/interface names to the corresponding wrapper
+class above"""
+
+# Importing this module will result in the dictionary above
+# being imported. E.g.,
+# >>> from pennylane.tape.transforms functions as fn
+# >>> fn["np"]
+# <class NumPyFunctions>
+# >>> fn["np"].abs
+#sys.modules[__name__] = MLFunctionWrapper
```

Comment:
I think this created the circular import trouble.

Reply:
:+1: This was me playing around to see if I could get auto framework detection working 😆 I could never get it to work though, so this is fine to remove!